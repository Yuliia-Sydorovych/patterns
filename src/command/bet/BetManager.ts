import { IPlayer } from "../../observer/interface/IPlayer";

export class BetManager
{
  private player: IPlayer<number>;

  constructor(player: IPlayer<number>)
  {
    this.player = player;
  }

  public placeBet(amount: number): boolean
  {
    return this.player.placeBet(amount);
  }

  public updateBalance(payout: number): void
  {
    this.player.updateBalance(payout);
  }

  public checkBet(value: number): boolean
  {
    return this.player.checkBet(value);
  }
}
