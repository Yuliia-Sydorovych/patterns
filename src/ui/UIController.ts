import { IMediator } from "../mediator/interface/IMediator";
import { IUIController } from "./interface/IUIController";

export class UIController implements IUIController
{
    private mediator!: IMediator;
    private spinButton: HTMLButtonElement;
    private addFundsButton: HTMLButtonElement;
    private betButtons: NodeListOf<HTMLButtonElement>;
    private balanceElement: HTMLElement;
    private resultMessage: HTMLElement;
    private betAmount: number = 1;
  
    constructor()
    {
        this.spinButton = document.getElementById("spin-button") as HTMLButtonElement;
        this.addFundsButton = document.getElementById("add-funds-button") as HTMLButtonElement;
        this.betButtons = document.querySelectorAll(".bet-button") as NodeListOf<HTMLButtonElement>;
        this.balanceElement = document.getElementById("balance")!;
        this.resultMessage = document.getElementById("game-result")!;

        this.setupSpinButton();
        this.setupBetButtons();
        this.setupFundsButton();
    }

    public bindMediator(mediator: IMediator): void
    {
        this.mediator = mediator;
    }

    public updateGameResult(message: string): void
    {
        this.resultMessage.textContent = message;
    }

    public setSpinButtonState(enabled: boolean): void
    {
        this.spinButton.disabled = !enabled;
    }

    public setBetButtonsState(enabled: boolean): void
    {
        this.betButtons.forEach((button) =>
        {
            button.disabled = !enabled;
        });
    }

    public setBalanceValue(value: number): void
    {
        this.balanceElement.textContent = `Balance: $${value}`;
    }

    private setupSpinButton(): void
    {
        this.spinButton.addEventListener("click", () =>
        {
            this.mediator?.notify(this.spinButton, "spin", this.betAmount);

            this.setBetButtonsState(false);
            this.setSpinButtonState(false);
        });
    }

    private setupBetButtons(): void
    {
        this.betButtons.forEach((button) =>
        {
            button.addEventListener("click", () =>
            {
                this.selectBet(button);
            });
        });

        this.selectBet(this.betButtons[0]);
    }

    private selectBet(button: HTMLButtonElement): void
    {
        this.betButtons.forEach((btn) => btn.classList.remove("selected"));
        button.classList.add("selected");

        this.betAmount = parseInt(button.textContent!.replace('$', ''));
        (document.getElementById("bet-amount") as HTMLInputElement).value = this.betAmount.toString();

        this.mediator?.notify(button, "bet", this.betAmount);
    }

    private setupFundsButton(): void
    {
        this.addFundsButton.addEventListener("click", () =>
        {
            this.mediator?.notify(this.addFundsButton, "funds", 100);

            this.updateGameResult("You added $100. Spin to play!");
        });
    }
}
  