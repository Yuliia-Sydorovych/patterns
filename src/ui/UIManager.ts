import { IUIController } from "./interface/IUIController";

export class UIManager
{
    private uiController: IUIController;
  
    constructor(uiController: IUIController)
    {
      this.uiController = uiController;
    }
  
    public updateOnSpinStart(): void
    {
      this.uiController.setSpinButtonState(false);
      this.uiController.setBetButtonsState(false);
      this.uiController.updateGameResult("Spinning...");
    }
  
    public updateOnSpinEnd(payout: number): void
    {
      this.uiController.setSpinButtonState(true);
      this.uiController.setBetButtonsState(true);
  
      const message = payout > 0 ? `You won $${payout}!` : "You lost this round.";
      this.uiController.updateGameResult(message);
    }
  
    public showInsufficientFunds(): void
    {
      this.uiController.updateGameResult("Insufficient funds to place the bet.");
    }

    public updateGameResult(value: string): void
    {
      this.uiController.updateGameResult(value);
    }
  }