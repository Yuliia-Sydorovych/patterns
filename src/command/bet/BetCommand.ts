import { UIManager } from "../../ui/UIManager";
import { ICommand } from "../interface/ICommand";
import { BetManager } from "./BetManager";

export class BetCommand implements ICommand
{
  private betManager: BetManager;
  private uiManager: UIManager;

  constructor(betManager: BetManager, uiManager: UIManager)
  {
    this.betManager = betManager;
    this.uiManager = uiManager;
  }

  public execute(value: number): void
  {
    const result = this.betManager.checkBet(value) ? `Bet of $${value} placed successfully.` : "Insufficient funds to place the bet.";
    this.uiManager.updateGameResult(result);
  }
}
