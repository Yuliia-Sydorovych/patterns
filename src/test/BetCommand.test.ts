import { BetCommand } from "../command/bet/BetCommand";
import { BetManager } from "../command/bet/BetManager";
import { UIManager } from "../ui/UIManager";

describe("BetCommand", () =>
{
  let betCommand: BetCommand;
  let betManager: BetManager;
  let uiManager: UIManager;

  beforeEach(() =>
  {
    betManager = { checkBet: jest.fn(), updateBalance: jest.fn() } as any;
    uiManager = { updateGameResult: jest.fn() } as any;
    betCommand = new BetCommand(betManager, uiManager);
  });

  it("should place a bet successfully if funds are sufficient", () =>
  {
    (betManager.checkBet as jest.Mock).mockReturnValue(true);

    betCommand.execute(100);

    expect(uiManager.updateGameResult).toHaveBeenCalledWith("Bet of $100 placed successfully.");
    expect(uiManager.updateGameResult).toMatchSnapshot();
  });

  it("should fail to place bet if funds are insufficient", () =>
  {
    (betManager.checkBet as jest.Mock).mockReturnValue(false);

    betCommand.execute(100);

    expect(uiManager.updateGameResult).toHaveBeenCalledWith("Insufficient funds to place the bet.");
    expect(uiManager.updateGameResult).toMatchSnapshot();
  });

  it("should undo the last bet", () =>
  {
    betCommand["lastBet"] = 100;
    betCommand.undo();

    expect(betManager.updateBalance).toHaveBeenCalledWith(100);
    expect(uiManager.updateGameResult).toHaveBeenCalledWith("Undo bet of $100");
    expect(uiManager.updateGameResult).toMatchSnapshot();
  });

  it("should redo the last bet", () =>
  {
    betCommand["lastBet"] = 50;
    jest.spyOn(betCommand, "execute");

    betCommand.redo();

    expect(betCommand.execute).toHaveBeenCalledWith(50);
    expect(uiManager.updateGameResult).toMatchSnapshot();
  });
});
