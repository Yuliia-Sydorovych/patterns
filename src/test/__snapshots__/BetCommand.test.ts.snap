// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BetCommand should fail to place bet if funds are insufficient 1`] = `
[MockFunction] {
  "calls": [
    [
      "Insufficient funds to place the bet.",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`BetCommand should place a bet successfully if funds are sufficient 1`] = `
[MockFunction] {
  "calls": [
    [
      "Bet of $100 placed successfully.",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`BetCommand should redo the last bet 1`] = `
[MockFunction] {
  "calls": [
    [
      "Insufficient funds to place the bet.",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`BetCommand should undo the last bet 1`] = `
[MockFunction] {
  "calls": [
    [
      "Undo bet of $100",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
